pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+				}
+
+				if retryDuration != doNotRetry {
+					//Retry duration should match
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		obtainedErr, retryDuration := controller.processServiceDeletion("external-balancer")
+		if err := tst.checkExpectedFn(obtainedErr, retryDuration); err != nil {
+			t.Errorf("%v processServiceDeletion() %v", tst.testName, err)
+		}
+	}
+
+}
+
+// Add unit testing for needsUpdate functions
+
+func TestDoesExternalLoadBalancerNeedsUpdate(t *testing.T) {
+
+	var oldSrv, newSrv *v1.Service
+
+	testCases := []struct {
+		testName       string //Name of the test case
+		updateFn       func() //Function to update the old and new service varuables
+		expectedResult bool   //needsupdate always returns bool
+
+	}{
+		{
+			testName: "If the service type is different",
+			updateFn: func() {
+				oldSrv = defaultExternalService()
+				newSrv = defaultExternalService()
+				newSrv.Spec.Type = v1.ServiceTypeClusterIP
+			},
+			expectedResult: true,
+		},
+		{
+			testName: "If the Ports are different",
+			updateFn: func() {
+				oldSrv = defaultExternalService()
+				newSrv = defaultExternalService()
+				oldSrv.Spec.Ports = []v1.ServicePort{
+					{
+						Port: 8000,
+					},
+					{
+						Port: 9000,
+					},
+					{
+						Port: 10000,
+					},
+				}
+				newSrv.Spec.Ports = []v1.ServicePort{
+					{
+						Port: 8001,
+					},
+					{
+						Port: 9001,
+					},
+					{
+						Port: 10001,
+					},
+				}
+
+			},
+			expectedResult: true,
+		},
+		{
+			testName: "If externel ip counts are different",
+			updateFn: func() {
+				oldSrv = defaultExternalService()
+				newSrv = defaultExternalService()
+				oldSrv.Spec.ExternalIPs = []string{"old.IP.1"}
+				newSrv.Spec.ExternalIPs = []string{"new.IP.1", "new.IP.2"}
+			},
+			expectedResult: true,
+		},
+		{
+			testName: "If externel ips are different",
+			updateFn: func() {
+				oldSrv = defaultExternalService()
+				newSrv = defaultExternalService()
+				oldSrv.Spec.ExternalIPs = []string{"old.IP.1", "old.IP.2"}
+				newSrv.Spec.ExternalIPs = []string{"new.IP.1", "new.IP.2"}
+			},
+			expectedResult: true,
+		},
+		{
+			testName: "If UID is different",
+			updateFn: func() {
+				oldSrv = defaultExternalService()
+				newSrv = defaultExternalService()
+				oldSrv.UID = types.UID("UID old")
+				newSrv.UID = types.UID("UID new")
+			},
+			expectedResult: true,
+		},
+	}
+
+	controller, _, _ := newController()
+	for _, tst := range testCases {
+		tst.updateFn()
+		obtainedResult := controller.needsUpdate(oldSrv, newSrv)
+		if obtainedResult != tst.expectedResult {
+			t.Errorf("%v needsUpdate() should have returned %v but returned %v", tst.testName, tst.expectedResult, obtainedResult)
+		}
+	}
+}
+
+func TestServiceCache(t *testing.T) {
+
+	sC := &serviceCache{serviceMap: make(map[string]*cachedService)} //ServiceCache
```

Comment:
All the testcases will be sharing this one cache, is this intended? And doe the order of test cases matter? In that case, I think we should at least put a comment so folks won't break it.